### DISCUSSION TITLE
Moving Selection Box to a toggle/different tool

### DISCUSSION THREAD

**Discussion by dotremilie** [ppy/osu#29886] [Ideas]
Hi, I wanted to discuss moving the Selection Box to a separate tool from Selection. This means selection alone would only selects the objects. This is one of things that makes me not consider Lazer for mapping. As a person who makes maps for 7 years already I have a couple of reasons to not see the box every time I select objects.
- It obstructs vision, so it's harder to focus and make even visual spacing when moving objects that were already placed
- I know all the shortcuts, so I will never use the the buttons that replicate CTRL+G/H/J/</> etc. behavior
- I don't always need to transform the selection, so always seeing the handles for transformation is obnoxious
- I make any scale or rotation transform with the tools under CTRL+SHIFT+R/S (on stable) because it's faster when you have muscle memory than with handles

As I use Photoshop a lot I see this tool as the same as CTRL+T vs Selection. Selection let's you select the layer and once you click CTRL+T you see the handles for transformation.

![obraz](https://github.com/user-attachments/assets/8c45dfe9-3268-4188-a559-b271707dd0ff)

 There is a couple of things to be discussed if people like the idea though. 

**Which toolbox should it be in? Toggles, because it's a toggle, Transform, because it's a transform tool or moved to View, to let user decide to either always show it or not?**

Personally the last option is one that I wouldn't even consider, because what if I actually want to transform something quickly with the handles. The other options are 50/50, they just decide on the behavior of the tool, because if it's in Toggles it should behave the same as Distance/Grid snap - once toggled it is toggled. If it's in Transform then it should behave similarly to other tools in there - if there is no selection it should be disabled, deselecting resets the state and you can only change it's state when object is selected. This one feels more correct than the other, but removes the option to always have the selection box visible and it would suck for Lazer-only mappers (mappers that started mapping on Lazer) so I'm more towards Toggles toolbox.

I already made the changes on my fork but wanted to discuss them first before I make a PR. This is how it would look:

https://github.com/user-attachments/assets/9b6d6e33-58d4-489b-8b37-9d850254d9a4

**Comment by bdach** [comment]
I'm not super sure we want to consider this, but I would say that the current behaviour of showing the selection box *should be the default behaviour for all users who are new to the editor*.

I have generally not seen many people complain about the editor handles until this point, so from my viewpoint this is a problem that is only felt by a small class of power users who have grown accustomed to using the hotkeys - but the thing is that *we cannot expect newcomers to the editor to be at that level already*. So I'd say that the default behaviours should be to have the selection box controls, and for you and others who have however many years' experience with the editor and don't need them, there should be an opt-*out* option.

That said please don't treat what I say as an official opinion because I'd want to hear @peppy's thoughts first.

**Reply by dotremilie** [reply]
Hi, thanks for an answer. That's why I made the discussion in the first place, obviously more input is needed.

Personally, If I wasn't actively mapping I'd already fully transition to lazer at this point. Editor is a big no in the current state. That's because the gameplay changes are so little compared to a whole change of workflow in lazer. You may not see the the feedback on it because people see lazer editor as a cool thing to check, rather than a tool that they'd use. The amount of changes is just not worth learning, therefore people don't even bother.

The handles look fine at first, and that was my first thought when I was only trying out some things, so from a developer viewpoint it may not seem like an issue. When I wanted to map in the style I usually do I never used them, because my mapping is very shortcut heavy.

**Reply by OliBomby** [reply]
I'm with @bdach on this one. I think the selection box is a great tool for new mappers and it should be the default. The way it shows up and immediately visually explains itself makes it easily discoverable by even someone randomly clicking around.

I'm also a mapper coming from stable and I didn't use the selection box that much because I know all the shortcuts. I think you either use it or you don't, so a toggle somewhere to turn it off would be nice. I'd put this toggle in the view menu, because it doesn't make sense to toggle this often and it's not worth shifting the hotkeys for grid snap and distance snap.

**Reply by minetoblend** [reply]
Hi, since you mentioned not having seen many people complain about it I figured I'd give my two cents, since I also want to be able to disable it.
Here's a list of reasons why I want to hide the box, but also what I think could be improved about it:

- 90% of the time I don't wanna scale/rotate my selection except for moving stuff around or ctrl+h, ctrl+j, etc. So the box is usually just visual noise.
- I often misclick the buttons underneath/above the select box when I actually wanna click what's underneath it.
- Even if I want to rotate/scale stuff, the dedicated scale/rotate tools are preferrable for me. Those having dedicated UI in the playfield for scaling/rotating would be very nice though, however I generally prefer gizmos over a box with handles for transforming anything that isn't an image, like this:

https://github.com/user-attachments/assets/3fe9d410-4128-4462-a4ae-f04d21ede42b

That being said I get that the box is more beginner friendly, this is just a preference of mine. My background in 3d software probably plays a big role in that, since pretty much all of them use gizmos for transformation. I think it's worth noting however that nearly every single one of those programs seems to arrive at the same exact component for this, so it seems to be doing a pretty good job for them.

- I usually wanna scale around the selection's center, which the box can't (currently) do.
- I also sometimes wanna scale around arbitrary points, which is impossible with the box.
- For some reason, it uses the handle I'm dragging as the scale origin. This is pretty much the opposite of any select box I've ever used in any other software. It turns what would just be "Drag point A to point B" into kind of into a guessing game where I need to make a bunch of micro-adjustments to make it do what I actually want.
- Including the hitobjects's radius for scaling makes no sense conceptually. I'm actually scaling the object's centers so really the box should look like this:
![image](https://github.com/user-attachments/assets/e0bb0377-7ba2-4681-8837-27833bc96c0e)
Obviously that looks bad and is confusing, which is why I think the box is fundamentally the wrong choice for scaling anything with a radius. This also comes with the issue that the box has a minimum width that it can't go below, which feels a bit awkward.
- I can't use it to scale to negative values.
- If I want to rotate the selection by 180 degrees, I wanna do that by clicking the CW/CCW button twice, however the box changes bounds with each click and the second click will go to whatever is underneath the box.
- Rotation feels awkward because the box isn't actually rotating.

If some of this stuff goes out of scope for this post I'm happy to ahead and make a new post for those things.

**Reply by OliBomby** [reply]
Yea sure dude, we can add the gizmos after we opt-out of the selection box

**Reply by peppy** [reply]
To answer @bdach here, I also hope we don't need to move this or make it a toggle. I'd sooner be looking for ways to improve what we have, reducing the visual clutter by making it fade into the background when not being interacted with, or hiding when not relevant.

Some of the points made above regarding the usability of the selection area stand, and I believe we should look to fix those before moving it to another tool or disabling it.

**Reply by dotremilie** [reply]
I agree it's a cool tool for newcomers and people learning to map, but what about people who already spent hundreds/thousands of hours and _won't_ use the handles because they're unnecessary? And it's not a minority. Once the newcomers learn to use shortcuts they might not use them either. For me - it's a dealbreaker when it comes to mapping on lazer, as well as current state of hitsounding, object-tied SVs and too much stuff happening on hover, but that's for another post. Like @minetoblend said, there are better transform tools available than the free transform box and I think it should be _at least_ toggleable if you don't want to move it to separate tool. I never need to see the box and I believe other people may not need to see it as well. 

> reducing the visual clutter by making it fade into the background when not being interacted with, or hiding when not relevant.

You don't mean more hover states, right? It'd be even more annoying than it is already. A toggle does exactly what you said. Hiding it when not needed and showing when user wants - intuitive and straight-forward.

My main point on moving it entirely to a separate tool is that Selection is the tool that let's you select the objects. If you want to transform the selection you will use **Transform** tools. The box should never be a `SelectionBox` in the first place, but rather a `FreeTransformBox` or anything close to that. If the objects weren't highlighted already in another way and the box would not have handles for transformation I'd agree to call it `SelectionBox` as it's **only** purpose would be to _box_ the _selection_.

**Reply by peppy** [reply]
If it's going to be moved, it isn't going to be a toggle. It should be implemented like transform mode in graphics apps: You choose to enter transform move and it stays visible for the duration of that transform. And I'd hope it's not too hidden away (probably leave one button when a selection is made to enter transform mode).

The question then is whether this reduces the usefulness of the tool as you'd need to enter transform mode each time you make a new selection.

**Reply by dotremilie** [reply]
Well i didn't say it has to be necessarily a toggle, I even gave different options. Nonetheless, the behavior described is more or less always a toggle, because it's either displaying or not. I don't mind any form of implementation, for me it just can still be a `SelectionBox` (though I already mentioned that naming it this way is always misleading) with a visibility boolean and that's how I made it on the video. Just let users disable/enable it whenever they want, I'll take anything.
